### ISSUE TITLE
Evaluating loss on validation dataset for early stopping

### ISSUE THREAD

**Issue by arunppsg** [deepchem/deepchem#3433]
Problem: I need to early stop training of a deepchem model to avoid overfitting. To this end, I need to compute evaluation loss on the validation dataset but I am not sure what is the best way to do it.

```py
# snippet for training a model
import deepchem as dc
import numpy as np

featurizer = dc.feat.ConvMolFeaturizer()

smiles = ['CCC1=[O+][Cu-3]2([O+]=C(CC)C1)[O+]=C(CC)CC(CC)=[O+]2', 
          'C(=Cc1ccccc1)C1=[O+][Cu-3]2([O+]=C(C=Cc3ccccc3)CC(c3ccccc3)=[O+]2)[O+]=C(c2ccccc2)C1',
          'CC(=O)N1c2ccccc2Sc2c1ccc1ccccc21',
          'Nc1ccc(C=Cc2ccc(N)cc2S(=O)(=O)O)c(S(=O)(=O)O)c1']

train_dataset = dc.data.NumpyDataset(X=featurizer.featurize(smiles), y=np.array([[0], [1], [0], [1]]))

metrics = [dc.metrics.Metric(dc.metrics.roc_auc_score)]
loss = dc.models.losses.BinaryCrossEntropy()
model = dc.models.GraphConvModel(mode='classification', n_tasks=1)

training_loss_value = model.fit(train_dataset, nb_epoch=1)

valid_dataset = train_dataset
```

### Method 1: Using DeepChem's loss function
```py
import tensorflow as tf

eval_preds = model.predict(train_dataset)
y = valid_dataset.y

temp = np.zeros((y.size, y.max() + 1))
temp[np.arange(y.size), y.squeeze()] = 1
y = temp

# Note: We have to know beforehand whether it is a tensorflow model or pytorch model for 
loss_value = loss._compute_tf_loss(eval_preds.squeeze(), y)
tf.add_n(loss_value).numpy()
```

### Method 2: Using BCELoss from PyTorch
```py
import torch
import torch.nn as nn

y = train_dataset.y.squeeze()
preds = eval_preds.squeeze()[np.arange(y.size), y]
loss_fn = nn.BCELoss(reduction='sum')
y = torch.from_numpy(y.astype('float32'))
preds = torch.from_numpy(preds)
loss_fn(preds, y)
```

I am wondering is there a better way to do it? Something like `model.eval_loss(valid_dataset)` will be helpful.
